feat: generate intercepted method body only once

Gh0stWalk3r opened the pull request:
resolves #9

Eliminates multiple generations of identical method bodies for intercepted methods and instead apply the `InterceptsLocation` attribute as many times as needed.

Gh0stWalk3r committed:
Fix FlashOWare/FlashOWare.Generators#9

Generate intercepted method body only once.
Apply InterceptsLocation attribute multiple times when method body is identical.

Gh0stWalk3r committed:
Use nameof() instead of self-written method name

Flash0ver commented on source/gen/FlashOWare.Generators/Text/Methods.cs:
```diff
@@ -2,6 +2,6 @@ namespace FlashOWare.Text;
 {
-	internal static string GetName { get; } = "GetName";
-	internal static string IsDefined { get; } = "IsDefined";
+	internal const string GetName  = nameof(Enum.GetName);
+	internal const string IsDefined = nameof(Enum.IsDefined);
```
I love the `nameof`-ification here.
And you're right, these are _constants_ ... the `System.Enum.GetName` method group won't change its name.

Flash0ver commented on source/gen/FlashOWare.Generators/Generators/Enumerations/EnumInterceptorGenerator.cs:
```diff
@@ -102,26 +102,26 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
 			document.Indent++;
 
-			IEnumerable<IGrouping<string, EnumMethodInvocation>> group = source.GroupBy(static value => value.Name);
+			IEnumerable<IGrouping<string, EnumMethodInvocation>> group = source.GroupBy(static value => $"{value.Name}:{value.Enum.Name}");
```
I love the fact that this change emits no duplicate code anymore. Really shows by removing the incrementing number suffix.
But utilizing _string-interpolation_ for the key-generation is not ideal, IMO.
_.NET Standard 2.0_ does not have the [DefaultInterpolatedStringHandler Struct](https://learn.microsoft.com/dotnet/api/system.runtime.compilerservices.defaultinterpolatedstringhandler). I'd like to avoid "expensive" string-concatenation here.

When writing this for the first _pre-release_, I actually thought about generating a ``ValueTuple`2``-based key here instead (via C# 7.0 tuple types), and utilizing its equality semantics here. And this Tuple could be given a descriptive name via a _using alias_ now with C# 12.0.

Of course benchmarking is in order to confirm that statement.
I'm currently building a Benchmarking-Library for Roslyn (incremental) Source Generators ... should be ready in 1 week.

Should this comment block this PR?
If you are happy to wait, I'd like to add Benchmarking first in the last week of November, and then get back to this PR and see which variant is really faster.
If you'd rather not delay completing this PR, I could also take a note in #13 to revisit this before publishing the next _pre-release_.

Gh0stWalk3r replied:
I am happy to wait with this PR until the Benchmark is added. In the meantime I will apply your suggestions 😄 

I absolutely agree, that the string-interpolation is not ideal and was more of a quick-shot solution to show what I mean with my issue report.

Flash0ver commented on source/gen/FlashOWare.Generators/Generators/Enumerations/EnumInterceptorGenerator.cs:
```diff
@@ -102,26 +102,26 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
-			IEnumerable<IGrouping<string, EnumMethodInvocation>> group = source.GroupBy(static value => value.Name);
+			IEnumerable<IGrouping<string, EnumMethodInvocation>> group = source.GroupBy(static value => $"{value.Name}:{value.Enum.Name}");
 
 			foreach (var invocations in group)
```
I made a mistake here ... since `group` is of type `IEnumerable<T>`, it should have a plural name instead: `groups`.
Would you mind fixing my mistake?

Also would you mind using explicit types over `var` for the following 3 `var`s? Also a previous mistake of mine.
In the `.editorconfig` I defnied `csharp_style_var_when_type_is_apparent = true:suggestion`.

(I'm reviewing my own code now and I am quite displeased 😉)

Flash0ver reviewed (changes requested):
Great improvement, especially when many method invocations are intercepted within a project.

Flash0ver committed:
Merge branch 'main' into pr/10

Flash0ver committed:
docs: update pre-release CHANGELOG

Flash0ver commented:
Merged from main (added documentation and CHANGELOGs) and added an entry to the Pre-Release CHANGELOG.

Gh0stWalk3r committed:
Apply feedback improvements

Use named ValueTuple `EnumInvocationGroup`
Rename variable `group` to `groups`
Use explicit type instead of var

Gh0stWalk3r committed:
Fix naming of parameter

Flash0ver commented on source/gen/FlashOWare.Generators/Generators/Enumerations/EnumInterceptorGenerator.cs:
```diff
@@ -8,6 +8,8 @@
 using FlashOWare.Text;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 
+using EnumInvocationGroup = (string methodName, string enumName);
```
Thank you for the improvements!
I like the name of the Tuple-Alias!

One final suggestion:
Since the generated Value-Tuple's fields are `public`, please begin both strings with an uppercase letter.

Other than that ... just waiting for #13 to complete and see an increase in performance ... and I will then merge this PR to include in the next _pre-release_.

Flash0ver commented on source/gen/FlashOWare.Generators/Generators/Enumerations/EnumInterceptorGenerator.cs:
```diff
@@ -16,7 +18,7 @@ public sealed class EnumInterceptorGenerator : IIncrementalGenerator
 	{
 		var source = context.SyntaxProvider.CreateSyntaxProvider(
-			static bool (SyntaxNode node, CancellationToken CancellationToken) =>
+			static bool (SyntaxNode node, CancellationToken cancellationToken) =>
```
❤️

Gh0stWalk3r committed:
Use PascalCase for public Value-Tuple fields

Flash0ver commented:
Thank you for your contribution! And sorry for the delay!
I'll complete #13 and  #14 this week (this time for real 😉) and will then merge this PR afterwards.

Flash0ver commented:
I am sorry for delaying this for so long. Now, at long last, it's mergin' time!

Flash0ver committed:
Merge branch 'main' into pr/10

Flash0ver committed:
test: add single interceptor per enum type

Flash0ver committed:
test: remove out-of-scope test case

Flash0ver reviewed (approved):
I added an entry to the pre-release changelog,
and - after (finally) completing #14 via PR #23 - added unit tests: all green.

Thank you for your contribution.
Thank you for your patience.
